File: src/test/java/com/vjiki/music/service/UserServiceIntegrationTest.java
package com.vjiki.music.service;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.catchThrowable;

import java.util.UUID;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;

import com.vjiki.music.dto.AuthRequest;
import com.vjiki.music.dto.AuthResponse;
import com.vjiki.music.dto.UserResponse;
import com.vjiki.music.entity.AccessLevel;
import com.vjiki.music.entity.AuthProvider;
import com.vjiki.music.entity.User;
import com.vjiki.music.repository.UserRepository;
import com.vjiki.music.support.AbstractIntegrationTest;

class UserServiceIntegrationTest extends AbstractIntegrationTest {

    @Autowired
    private UserServiceImpl userService;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    private User createTestUser(String email, String nickname) {
        return userRepository.save(
                User.builder()
                        .email(email)
                        .nickname(nickname)
                        .accessLevel(AccessLevel.USER)
                        .provider(AuthProvider.LOCAL)
                        .isActive(true)
                        .isVerified(false)
                        .createdBy("system")
                        .modifiedBy("system")
                        .build());
    }

    private User createTestUserWithPassword(String email, String nickname, String password) {
        return userRepository.save(
                User.builder()
                        .email(email)
                        .nickname(nickname)
                        .passwordHash(passwordEncoder.encode(password))
                        .accessLevel(AccessLevel.USER)
                        .provider(AuthProvider.LOCAL)
                        .isActive(true)
                        .isVerified(false)
                        .createdBy("system")
                        .modifiedBy("system")
                        .build());
    }

    @Nested
    @DisplayName("UserService Integration Tests")
    class UserServiceIntegrationTests {

        @Test
        @DisplayName("should get user by id")
        void shouldGetUserById() {
            User user = createTestUser("getuser@example.com", "getuser");

            UserResponse result = userService.getUserById(user.getId());

            assertThat(result.getId()).isEqualTo(user.getId());
            assertThat(result.getEmail()).isEqualTo("getuser@example.com");
            assertThat(result.getNickname()).isEqualTo("getuser");
        }

        @Test
        @DisplayName("should throw exception when user not found")
        void shouldThrowExceptionWhenUserNotFound() {
            Throwable exception = catchThrowable(() -> userService.getUserById(UUID.randomUUID()));

            assertThat(exception).isNotNull();
            assertThat(exception).hasMessageContaining("not found");
        }

        @Test
        @DisplayName("should authenticate user with valid credentials")
        void shouldAuthenticateUserWithValidCredentials() {
            String password = "password123";
            User user = createTestUserWithPassword("auth@example.com", "authuser", password);

            AuthResponse result = userService.authenticate(new AuthRequest("auth@example.com", password));

            assertThat(result.getAuthenticated()).isTrue();
            assertThat(result.getUserId()).isEqualTo(user.getId());
            assertThat(result.getMessage()).isEqualTo("Authentication successful");
        }

        @Test
        @DisplayName("should fail authentication with invalid password")
        void shouldFailAuthenticationWithInvalidPassword() {
            createTestUserWithPassword("invalid@example.com", "invaliduser", "correctpassword");

            AuthResponse result = userService.authenticate(new AuthRequest("invalid@example.com", "wrongpassword"));

            assertThat(result.getAuthenticated()).isFalse();
            assertThat(result.getMessage()).isEqualTo("Invalid email or password");
        }

        @Test
        @DisplayName("should fail authentication for inactive user")
        void shouldFailAuthenticationForInactiveUser() {
            String password = "password123";
            User user = createTestUserWithPassword("inactive@example.com", "inactiveuser", password);
            user.setIsActive(false);
            userRepository.save(user);

            AuthResponse result = userService.authenticate(new AuthRequest("inactive@example.com", password));

            assertThat(result.getAuthenticated()).isFalse();
            assertThat(result.getMessage()).isEqualTo("User account is inactive");
        }

        @Test
        @DisplayName("should fail authentication for non-existent user")
        void shouldFailAuthenticationForNonExistentUser() {
            AuthResponse result = userService.authenticate(new AuthRequest("nonexistent@example.com", "password"));

            assertThat(result.getAuthenticated()).isFalse();
            assertThat(result.getMessage()).isEqualTo("Invalid email or password");
        }
    }
}

File: src/test/kotlin/com/vjiki/music/service/UserServiceIntegrationTest.kt
package com.vjiki.music.service

import com.vjiki.music.config.TestContainersConfig
import com.vjiki.music.dto.AuthRequest
import com.vjiki.music.entity.AccessLevel
import com.vjiki.music.entity.AuthProvider
import com.vjiki.music.entity.User
import com.vjiki.music.repository.UserRepository
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.extensions.spring.SpringExtension
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.test.context.ContextConfiguration
import org.springframework.transaction.annotation.Transactional
import java.util.*

@SpringBootTest
@ContextConfiguration(classes = [TestContainersConfig::class])
@Transactional
open class UserServiceIntegrationTest : DescribeSpec() {

    override fun extensions() = listOf(SpringExtension)

    @Autowired
    lateinit var userService: UserService

    @Autowired
    lateinit var userRepository: UserRepository

    @Autowired
    lateinit var passwordEncoder: PasswordEncoder

    fun createTestUser(email: String, nickname: String): User {
        return userRepository.save(
            User(
                email = email,
                nickname = nickname,
                accessLevel = AccessLevel.USER,
                provider = AuthProvider.LOCAL,
                isActive = true,
                isVerified = false,
                createdBy = "system",
                modifiedBy = "system"
            )
        )
    }

    fun createTestUserWithPassword(email: String, nickname: String, password: String): User {
        val user = User(
            email = email,
            nickname = nickname,
            passwordHash = passwordEncoder.encode(password),
            accessLevel = AccessLevel.USER,
            provider = AuthProvider.LOCAL,
            isActive = true,
            isVerified = false,
            createdBy = "system",
            modifiedBy = "system"
        )
        return userRepository.save(user)
    }

    init {
        describe("UserService Integration Tests") {
        it("should get user by id") {
            val user = createTestUser("getuser@example.com", "getuser")
            val result = userService.getUserById(user.id)

            result.id shouldBe user.id
            result.email shouldBe "getuser@example.com"
            result.nickname shouldBe "getuser"
        }

        it("should throw exception when user not found") {
            val exception = kotlin.runCatching {
                userService.getUserById(UUID.randomUUID())
            }.exceptionOrNull()

            exception shouldNotBe null
            exception?.message?.contains("not found") shouldBe true
        }

        it("should authenticate user with valid credentials") {
            val password = "password123"
            val user = createTestUserWithPassword("auth@example.com", "authuser", password)

            val result = userService.authenticate(AuthRequest("auth@example.com", password))

            result.authenticated shouldBe true
            result.userId shouldBe user.id
            result.message shouldBe "Authentication successful"
        }

        it("should fail authentication with invalid password") {
            val user = createTestUserWithPassword("invalid@example.com", "invaliduser", "correctpassword")

            val result = userService.authenticate(AuthRequest("invalid@example.com", "wrongpassword"))

            result.authenticated shouldBe false
            result.message shouldBe "Invalid email or password"
        }

        it("should fail authentication for inactive user") {
            val password = "password123"
            val user = createTestUserWithPassword("inactive@example.com", "inactiveuser", password)
            user.isActive = false
            userRepository.save(user)

            val result = userService.authenticate(AuthRequest("inactive@example.com", password))

            result.authenticated shouldBe false
            result.message shouldBe "User account is inactive"
        }

        it("should fail authentication for non-existent user") {
            val result = userService.authenticate(AuthRequest("nonexistent@example.com", "password"))

            result.authenticated shouldBe false
            result.message shouldBe "Invalid email or password"
        }
    }
    }
}

